Content.Server/Discord/DiscordLink/DiscordChatLink.cs

```diff
@@ -0,0 +1,96 @@
+﻿using Content.Server.Chat.Managers;
+using Content.Shared.CCVar;
+using Content.Shared.Chat;
+using Discord.WebSocket;
+using Robust.Shared.Asynchronous;
+using Robust.Shared.Configuration;
+
+namespace Content.Server.Discord.DiscordLink;
+
+public sealed class DiscordChatLink
+{
+    [Dependency] private DiscordLink _discordLink = default!;
+    [Dependency] private IConfigurationManager _configurationManager = default!;
+    [Dependency] private readonly IChatManager _chatManager = default!;
+    [Dependency] private readonly ITaskManager _taskManager = default!;
+
+    private ulong? _oocChannelId;
+    private ulong? _adminChannelId;
+
+    public void Initialize()
+    {
+        _discordLink.OnMessageReceived += OnMessageReceived;
+
+        _configurationManager.OnValueChanged(CCVars.OocDiscordChannelId, OnOocChannelIdChanged, true);
+        _configurationManager.OnValueChanged(CCVars.AdminChatDiscordChannelId, OnAdminChannelIdChanged, true);
+    }
+
+    public void Shutdown()
+    {
+        _discordLink.OnMessageReceived -= OnMessageReceived;
+
+        _configurationManager.UnsubValueChanged(CCVars.OocDiscordChannelId, OnOocChannelIdChanged);
+        _configurationManager.UnsubValueChanged(CCVars.AdminChatDiscordChannelId, OnAdminChannelIdChanged);
+    }
+
+    private void OnOocChannelIdChanged(string channelId)
+    {
+        if (string.IsNullOrEmpty(channelId))
+        {
+            _oocChannelId = null;
+            return;
+        }
+
+        _oocChannelId = ulong.Parse(channelId);
+    }
+
+    private void OnAdminChannelIdChanged(string channelId)
+    {
+        if (string.IsNullOrEmpty(channelId))
+        {
+            _adminChannelId = null;
+            return;
+        }
+
+        _adminChannelId = ulong.Parse(channelId);
+    }
+
+    private void OnMessageReceived(SocketMessage message)
+    {
+        if (message.Author.IsBot)
+            return;
+
+        _taskManager.RunOnMainThread(() =>
+        {
+            if (message.Channel.Id == _oocChannelId)
+            {
+                _chatManager.SendHookOOC(message.Author.Username, message.Content);
+            }
+            else if (message.Channel.Id == _adminChannelId)
+            {
+                _chatManager.SendHookAdmin(message.Author.Username, message.Content);
+            }
```

Comment:
You should filter for channels outside the `RunOnMainThread()`, so that you're not bothering the main thread for every message in the Discord server.